### PR TITLE
liveness: allow registering callbacks after start

### LINKED ISSUE #106706: server: TestAddNewStoresToExistingNodes failed
server.TestAddNewStoresToExistingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10886615?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10886615?buildTab=artifacts#/) on master @ [665084ed379a25fcde92ee8b4c9dc48e192876e5](https://github.com/cockroachdb/cockroach/commits/665084ed379a25fcde92ee8b4c9dc48e192876e5):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 131489 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAddNewStoresToExistingNodes
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestAddNewStoresToExistingNodes363450179
    test_log_scope.go:81: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAddNewStoresToExistingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by tbg**
I discovered[^1] a deadlock scenario when multiple nodes in the cluster restart
with additional stores that need to be bootstrapped. In that case, liveness
must be running when the StoreIDs are allocated, but it is not.

Trying to address this problem, I realized that when an auxiliary Store is bootstrapped,
it will create a new replicateQueue, which will register a new callback into NodeLiveness.

But if liveness must be started at this point to fix #106706, we'll run into the assertion
that checks that we don't register callbacks on a started node liveness.

Something's got to give: we will allow registering callbacks at any given point
in time, and they'll get an initial set of notifications synchronously. I
audited the few users of RegisterCallback and this seems OK with all of them.

[^1]: https://github.com/cockroachdb/cockroach/issues/106706#issuecomment-1640254715

Epic: None
Release note (bug fix): it was possible for a node status to reflect a "last up" timestamp that lead the actual last liveness heartbeat of the node. This has been fixed.

### CODE REVIEW

**Comment by erikgrinaker** [comment]
Right, this is probably fallout from #103601.

**Comment by andrewbaptist** [comment]
Thanks Tobi!

I see what you mean about this test, and it was tricky to get some of tests, including this one, to work without doing what you did by allowing later registration. 

The approach you have works, but I had considered two other options as well (and could code them up if you want)
1) Don't allow disks/stores to be added to a running instance, instead require a restart. I wasn't sure if this was something that we support in a field deployment but I should have checked first. If we don't support this, then it would be possible to rearrange the code to allow this to work

2) Have a "single" node level callback that then fans out to the stores on liveness changes. This simplifies the liveness code as there is only a single callback, and the node already has locking in place for adding stores.

Your approach also works and is probably the most straightforward.

**Comment by tbg** [comment]
> Don't allow disks/stores to be added to a running instance, instead require a restart. I wasn't sure if this was something that we support in a field deployment but I should have checked first. If we don't support this, then it would be possible to rearrange the code to allow this to work

The new stores are being added following a restart. However, due to the need to assign StoreIDs, KV needs to be up and running. So these new stores are added "async". I don't see a good way around that.

> Have a "single" node level callback that then fans out to the stores on liveness changes. This simplifies the liveness code as there is only a single callback, and the node already has locking in place for adding stores.

That's not a bad idea, though I won't be able to do this until I leave (and it might be crowded out anyway). So even though I like that idea I'd like to stick with what I have, so that we can at least address the bug.

File: pkg/kv/kvserver/liveness/liveness.go [hunk 1]
```diff
@@ -641,13 +647,7 @@ func (nl *NodeLiveness) Start(ctx context.Context) {
...
 	// We may have received some liveness records from Gossip prior to Start being
 	// called. We need to go through and notify all the callers of them now.
-	for _, entry := range nl.ScanNodeVitalityFromCache() {
-		if entry.IsLive(livenesspb.IsAliveNotification) {
-			for _, fn := range nl.onIsLive {
-				fn(entry.GetInternalLiveness())
-			}
-		}
-	}
+	nl.notifyIsAliveCallbacks()
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/liveness/liveness.go:650]: This doesn't actually notify anything, it needs to pass in the callbacks, or the method needs to collect them. This also reveals lacking test coverage, can we add a quick test?

Comment (tbg) [reply pkg/kv/kvserver/liveness/liveness.go]: It's now tested, and I removed the footgun (need to pass a `[]IsLiveCallback`). This required some refactoring, basically picking up where @andrewbaptist left off at some point in the past (thanks for the initial leg work!)

File: pkg/kv/kvserver/liveness/liveness.go [hunk 2]
```diff
@@ -700,6 +687,8 @@ func (nl *NodeLiveness) Start(ctx context.Context) {
 					return nil
 				}); err != nil {
 				log.Warningf(ctx, heartbeatFailureLogFormat, err)
+			} else if nl.onSelfHeartbeat != nil {
+				nl.onSelfHeartbeat(ctx)
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/liveness/liveness.go:691]: This misses a bunch of cases. In particular, explicit `Heartbeat()` calls made during lease acquisitions.

We may want to move this into `heartbeatInternal` or something, or keep it in `updateLiveness` with the proper condition.

Comment (tbg) [reply pkg/kv/kvserver/liveness/liveness.go:691]: > This misses a bunch of cases. In particular, explicit Heartbeat() calls made during lease acquisitions.

This was intentional, see commit message. There is no need for "every" heartbeat to bump this timestamp. But we do want the last up timestamp to rise along at the pace of the heartbeat loop, which this code achieves.

The last up timestamp is used only to inform the "node up" event that is emitted to the event log. There is no way to record the "true" last time the node is up anyway (what does that even mean?), so this mechanism keeps it simple and gets the job done.

Please let me know if you think I'm missing something here.

Comment (andrewbaptist) [reply pkg/kv/kvserver/liveness/liveness.go:691]: This is definitely a bug before. The `onSelfHeartbeat` call should not be made for other nodes. Thanks for catching and fixing it!

**Review by andrewbaptist — APPROVED** [review]
:lgtm:

File: pkg/kv/kvserver/liveness/liveness.go (hunk 2 above)
Comment (erikgrinaker) [reply pkg/kv/kvserver/liveness/liveness.go:691]: > The last up timestamp is used only to inform the "node up" event that is emitted to the event log.

Oh right, that thing. I'd be inclined to just remove that outright, but yeah, let's leave that for another day.

File: pkg/kv/kvserver/liveness/storage.go [hunk 3]
```diff
@@ -24,15 +24,32 @@ import (
...
+	Get(ctx context.Context, nodeID roachpb.NodeID) (Record, error)
+	Update(
+		ctx context.Context, update LivenessUpdate, handleCondFailed func(actual Record) error,
+	) (Record, error)
+	Create(ctx context.Context, nodeID roachpb.NodeID) error
+	Scan(ctx context.Context) ([]Record, error)
+}
+
+// storageImpl implements Storage by storing entries in the replicated KV liveness range.
+type storageImpl struct {
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/liveness/storage.go:39]: nit: let's rename this `kvStorage`, c.f. `NewKVStorage`.

**Review by erikgrinaker — APPROVED** [review]

**Comment by erikgrinaker** [comment]
We'll want a targeted backport here too, since it seems like prior versions are vulnerable to the same issue (but I'm not sure with all the refactoring in this area).

**Comment by tbg** [comment]
> We'll want a targeted backport here too, since it seems like prior versions are vulnerable to the same issue (but I'm not sure with all the refactoring in this area).

I'm not sure we should fix this in a backport. The chances of messing something up are much higher than anyone hitting this bug. Folks very very rarely add stores to existing nodes (multi-storage in itself is rare) and doing so in a way that loses quorum on the storeIDGen is even rarer. Also, there is a workaround - restart a quorum without the extra store.

If you disagree, let's file an issue because I am not sure I would get to this backport even if we tried.

**Comment by erikgrinaker** [comment]
Ok, I can buy that. Was thinking of surrounding issues like the self callback too though, but they're fairly minor.

**Comment by tbg** [comment]
The self-callback I can fix in a backport - we can just add a conditional (if we don't insist on also adding a test).

The hang at startup, btw, I think has existed in every past version of CRDB, and to the best of my knowledge was never hit by anyone in a real cluster. Even the test that caught it only did so after I refactored it and made it use in-mem engines. So I am fairly confident that _not_ backporting is the reasonable strategy here.

**Comment by tbg** [comment]
bors r=erikgrinaker

**Comment by tbg** [comment]
Backports:


https://github.com/cockroachdb/cockroach/pull/107605
https://github.com/cockroachdb/cockroach/pull/107606

**Pull request merged**